### PR TITLE
fix(codegen): abstract effect not propagated to union siblings

### LINKED ISSUE #2492: bug: codegen generates incorrect types for sealed trait with lazy directive and abstractEffectType enabled
```graphql
directive @lazy on FIELD_DEFINITION

schema {
  query: Query
}

type Query {
    u: U
}


type A {
  field: Int @lazy
}

type B {
  field: Int
}

union U = A | B


```

```scala
import Types._

object Types {

  final case class A[F[_]](field: F[scala.Option[Int]]) extends U
  final case class B(field: scala.Option[Int]) extends U

  sealed trait U extends scala.Product with scala.Serializable

}

object Operations {

  final case class Query[F[_]](
      u: F[scala.Option[U[F]]]
                     // ^ Types.U does not take type parameters
  )

}
```

This is probably because SchemaWriter does not properly handle union that contains types with `F[_]` here.

https://github.com/ghostdogpr/caliban/blob/636b2bd1884e9a209ca12ce0bff88f4b0c98df49/tools/src/main/scala/caliban/tools/SchemaWriter.scala#L185

### PR DESCRIPTION
**Pull request by i10416**
fix https://github.com/ghostdogpr/caliban/issues/2492

### CODE REVIEW

**Comment by i10416** [comment]
@ghostdogpr Hello, thank you for maintaining this project. Could you be kind to review this change?

File: tools/src/test/scala/caliban/tools/SchemaWriterSpec.scala [hunk 1]
```diff
@@ -80,6 +80,32 @@ object SchemaWriterSpec extends SnapshotTest {
...
+           lazyField: Int @lazy
+         }
+         type B {
+           lazyFieldWithArgs(int: Int): Int @lazy
+         }
+         type C {
+           field: Int
+         }
+         type D {
+          fiend: Int
```
Comment (ghostdogpr) [inline tools/src/test/scala/caliban/tools/SchemaWriterSpec.scala]: it's very minor but there's a small typo 😄

**Review by ghostdogpr — APPROVED** [review]
Looks good!

File: tools/src/test/scala/caliban/tools/SchemaWriterSpec.scala (hunk 1 above)
Comment (i10416) [reply tools/src/test/scala/caliban/tools/SchemaWriterSpec.scala]: Thanks. I've just pushed the fix.

**Comment by ghostdogpr** [comment]
Thanks for fixing this! Not all edge cases and combinations have been tested. Will merge when the CI is green!

**Comment by i10416** [comment]
ref:  https://github.com/ghostdogpr/caliban/pull/2064
